File: NumOfIslands.cpp
// Time Complexity : O(n*m) as we are travellig throught all the elements in the given matrix.
// Space Complexity : O(1) if we do not consider recursive stack. o(n*m) with recursive stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : no issues as of now.. Learning



class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
       if(grid.size() == 0) return 0;
        
        int count = 0;
        
        for (int i = 0; i <grid.size() ; i++){
            
            for ( int j = 0; j < grid[0].size(); j++){// traversing throught the matrix for all the possible 1's i.e. islands
                if (grid[i][j] == '1'){
                    
                    recurr(grid,i,j); // recursion
                    count +=1; // count of posible islands
                }
            }
        }
        return count;
    }
    
    vector<vector<int>> directions = {{0,1},{0,-1},{1,0},{-1,0}};
    
    void recurr(vector<vector<char>>& grid , int i, int j){
        
        grid[i][j] = '0'; // updating the travelled island to 0 inorder to make sure we do not travel again
        

        for(vector<int> dir:directions){
            
            int r = i + dir[0];
            int c = j + dir[1];
            
            
            if( r>=0 && r<grid.size() && c >=0 && c <grid[0].size() && grid[r][c] == '1'){ // checking out of bounds and also the neighbour island.
                recurr(grid,r,c); // recursion
            }
        }
        
        
    }
};

File: Num of Islands.py
#  Time Complexity : O(nm)  - > 2mn
#  Space Complexity : O(mn)  
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : No 

# Use DFS
# Iterate through the whole matrix
# find if grid[i][j]==1
# go into DFS make grid[i][j]=#
# go into all 4 directions of i,j and check same

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if grid == None or len(grid)==0:
            return 0
        count=0
        
        directions = [(1,0),(0,1),(-1,0),(0,-1)]
        def dfs(r,c):
            #base
            if(r<0 or c<0 or r>=len(grid) or c>=len(grid[0]) or grid[r][c]=="#" or grid[r][c]=="0"):
                return
            
            #logic
            grid[r][c]="#"
            for dir in directions:
                rd = dir[0]+r
                cd = dir[1]+c
                if((rd>=0 and cd>=0 and rd<len(grid) and cd<len(grid[0])) and grid[rd][cd]=="1"):
                    dfs(rd,cd)
                    
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j]=="1":
                    count+=1
                    dfs(i,j)
        
        return count
                    
        